File: framework.py
#!/usr/bin/env python
"""
Nicholas' Example API code for interacting with Alienvault API.

This is just Example code written by NMA.IO.

There isn't really much you can do with the API just yet, so
this will be a work in progress.
Grab your API key here:
    https://www.alienvault.com/documentation/usm-anywhere/api/alienvault-api.htm?cshid=1182

That said, you could use this to write an alerter bot.

* Note, this isn't an SDK, just some example code to get  people started...

"""
# import json  # uncomment if you want pretty printing.
import base64
import requests

URL = "alienvault.cloud/api/2.0"
HOST = ""  # put your subdomain here.


def Auth(apiuser, apikey):
    """Our Authentication Code.

    :params apiuser
    :params apikey

    :returns oauth_token
    """
    headers = {"Authorization": "Basic {}".format(base64.b64encode("%s:%s" % (apiuser, apikey)))}

    r = requests.post("https://{}.{}/oauth/token?grant_type=client_credentials".format(HOST, URL), headers=headers)
    if r.status_code is 200:
        return r.json()["access_token"]
    else:
        print("Authentication failed. Check username/Password")
        exit(1)


def Alarms(token):
    """Pull Alarms from the API Console."""
    headers = {"Authorization": "Bearer {}".format(token)}
    r = requests.get("https://{}.{}/alarms/?page=1&size=20&suppressed=false&status=open".format(HOST, URL), headers=headers)
    if r.status_code is not 200:
        print("Something went wrong. \n{}".format(r.content))
        exit(1)
    else: return (r.json())


def Events():
    """Nothing yet."""
    pass

if __name__ == "__main__":
    print("Simple API Integration with Alienvault USM Anywhere - 2018 NMA.IO")
    token = Auth("username", "password")
    jdata = Alarms(token)
    for item in jdata["_embedded"]["alarms"]:
        # print(json.dumps(item, indent=2)) # uncomment if you want a pretty version of the whole block.
        print(item["rule_method"] + ": " + " ".join(item["alarm_sources"]))

File: framework.go
package main

/*
Quick Alienvault USM Anywhere client for Go.

Requires gjson for json parsing:
	go get github.com/tidwall/gjson

NMA 2018.
*/
import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"strings"

	"github.com/tidwall/gjson"
)

var URL string = "alienvault.cloud/api/2.0"
var HOST string = "yoursubdomain"

func Auth(apiUser, apiKey string) (token string) {
	client := &http.Client{}
	v := url.Values{}
	apilink := fmt.Sprintf("https://" + HOST + "." + URL + "/oauth/token?grant_type=client_credentials")
	method := "POST"
	req, err := http.NewRequest(method, apilink, strings.NewReader(v.Encode()))

	req.SetBasicAuth(apiUser, apiKey)
	resp, err := client.Do(req)
	if err != nil {
		log.Println(err)
		log.Fatal("Alienvault OAuth Error.")
		return
	}
	defer resp.Body.Close()
	json, _ := ioutil.ReadAll(resp.Body)

	if strings.Contains(string(json), "Unauthorized") {
		log.Fatal(string(json)) // Bad API Key
		return
	}
	token = gjson.Get(string(json), "access_token").String()
	if len(token) < 128 {
		log.Fatal("No Token Received.") // We shouldn't get here.
	}
	return
}

func Alarms(accessToken string) string {
	client := &http.Client{}
	v := url.Values{}
	apilink := fmt.Sprintf("https://" + HOST + "." + URL + "/alarms/?page=1&size=20&suppressed=false&status=open")
	method := "GET"
	req, err := http.NewRequest(method, apilink, strings.NewReader(v.Encode()))
	req.Header.Set("Authorization", "Bearer "+accessToken)
	resp, err := client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()
	json, _ := ioutil.ReadAll(resp.Body)
	jsonOutput := gjson.Get(string(json), "_embedded.alarms").String()

	return jsonOutput
}

func Events() {
	// Not implemented ... But here is where you can get more data.
}

func main() {
	key := Auth("apiclientname", "secret")
	x := Alarms(key)
	rule := gjson.Get(x, "#.rule_method")
	src := gjson.Get(x, "#.alarm_sources")
	for i := range rule.Array() {
		fmt.Println(rule.Array()[i], src.Array()[i])
	}

}
